File: 프로그래머스/2/132265. 롤케이크 자르기/롤케이크 자르기.py
from collections import Counter


def solution(topping):
    answer = 0
    
    cnt = Counter(topping)
    check = set()
    
    for t in topping:
        cnt[t] -= 1
        if cnt[t] == 0:
            cnt.pop(t)
        
        check.add(t)
        answer += len(check) == len(cnt)
    
    return answer

File: 프로그래머스/2/132265. 롤케이크 자르기/롤케이크 자르기.java
class Solution {
    
    public int solution(int[] topping) {
        int answer = 0;

        int max = 10000;
        int[] count = new int[max + 1];

        int rightKinds = 0;
        for (int t : topping) {
            if (count[t]++ == 0) rightKinds++;
        }

        boolean[] leftVisited = new boolean[max + 1];
        int leftKinds = 0;

        for (int t : topping) {
            if (leftVisited[t] == false) {
                leftVisited[t] = true;
                leftKinds++;
            }

            if (--count[t] == 0) {
                rightKinds--;
            }

            if (leftKinds == rightKinds) {
                answer++;
            }
        }

        return answer;
    }
}